### DISCUSSION TITLE
Is it possible to upload multiple files?

### DISCUSSION THREAD

**Discussion by branislav-brincko** [honojs/hono#3384] [General]
Docs talk about uploading a SINGLE file:

![image](https://github.com/user-attachments/assets/36d45a20-eecb-4cfb-9399-99bca8dcc866)

Is it possible to upload multiple files though? My frontend code:

```html
<!doctype html>
<html lang="en">
	<head>
		<meta charset="UTF-8" />
		<meta name="viewport" content="width=device-width, initial-scale=1.0" />
		<title>File Upload</title>
	</head>
	<body>
		<form id="uploadForm" enctype="multipart/form-data">
			<input type="file" id="files" name="files" multiple />
			<button type="button" onclick="uploadFiles()">Upload</button>
		</form>

		<script>
			function uploadFiles() {
				const filesInput = document.getElementById('files')
				const formData = new FormData()

				// Append all selected files to the formData
				for (let i = 0; i < filesInput.files.length; i++) {
					formData.append('file', filesInput.files[i])
				}

				// Use Fetch API to send files to the backend
				fetch('http://localhost:3000/api/forms/test-form', {
					method: 'POST',
					body: formData,
				})
					.then((response) => response.json())
					.then((data) => console.log(data))
					.catch((error) => console.error('Error:', error))
			}
		</script>
	</body>
</html>
```

Hono:

```ts
	.post('/test-form', async (c) => {
		const body = await c.req.parseBody()
		const { files } = body
		console.log('files', files)
		return c.json({ ok: 'ok' })
	})
```

The logged "files" property is always a single file. 

![image](https://github.com/user-attachments/assets/1c219d44-980c-42a5-8aa3-9e9fc9698199)

Can we handle mutliple files somehow? 

Thanks 🙏

**Comment by branislavbrincko** [comment]
I found the answer - https://hono.dev/docs/api/request#multiple-files-with-same-name

**Comment by Tommy1223454** [comment]
Once multiple-file upload enters the picture, I’d be extra careful not to treat it as just a transport problem.

The bigger risk is usually that each file is accepted as-is without a dedicated server-side inspection step before storage or further processing.

I built an OSS package for this in Node.js:
https://github.com/pompelmi/pompelmi

It’s useful when you want upload handling to include actual trust checks instead of only multipart parsing.
